dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java

```diff
@@ -0,0 +1,234 @@
+            limitedChan = concurrencyLimiter(limitedChan, conf, clientMetrics, clock);
+            return new FixedLimitedChannel(limitedChan, MAX_REQUESTS_PER_CHANNEL, clientMetrics);
+        }
+
+        private static Channel retryingChannel(
+                Channel channel,
+                ClientConfiguration conf,
+                Supplier<ListeningScheduledExecutorService> scheduler,
+                Random random) {
+            if (conf.maxNumRetries() == 0) {
+                return channel;
+            }
+
+            return new RetryingChannel(
+                    channel,
+                    conf.maxNumRetries(),
+                    conf.backoffSlotSize(),
+                    conf.serverQoS(),
+                    conf.retryOnTimeout(),
+                    scheduler.get(),
+                    random::nextDouble);
+        }
+
+        private static LimitedChannel concurrencyLimiter(
+                LimitedChannel channel, ClientConfiguration config, DialogueClientMetrics metrics, Ticker clock) {
+            ClientConfiguration.ClientQoS clientQoS = config.clientQoS();
+            switch (clientQoS) {
+                case ENABLED:
+                    return new ConcurrencyLimitedChannel(
+                            channel, ConcurrencyLimitedChannel.createLimiter(clock), metrics);
+                case DANGEROUS_DISABLE_SYMPATHETIC_CLIENT_QOS:
+                    return channel;
+            }
+            throw new SafeIllegalStateException(
+                    "Encountered unknown client QoS configuration", SafeArg.of("ClientQoS", clientQoS));
+        }
+
+        private static LimitedChannel nodeSelectionStrategy(
+                List<LimitedChannel> channels,
+                VersionedTaggedMetricRegistry metrics,
+                ClientConfiguration config,
+                Optional<DialogueChannel> maybeExisting,
+                Random random) {
+
+            if (channels.size() == 1) {
+                return channels.get(0); // no fancy node selection heuristic can save us if our one node goes down
+            }
+
+            switch (config.nodeSelectionStrategy()) {
+                case PIN_UNTIL_ERROR:
+                case PIN_UNTIL_ERROR_WITHOUT_RESHUFFLE:
+                    // create a fresh instance (ignoring the idea of live reloading initially) - we might discard this
+                    PinUntilErrorChannel goalInstance = PinUntilErrorChannel.of(
+                            config.nodeSelectionStrategy(), channels, DialoguePinuntilerrorMetrics.of(metrics), random);
+
+                    // find out whether we could have live-reloaded to get to the goal state
+                    Optional<PinUntilErrorChannel> maybeLiveReloaded = maybeExisting
+                            .flatMap(existing -> goalInstance.reloadableFrom(existing.nodeSelectionStrategy))
+                            .map(existing -> existing.liveReloadNewInstance(channels));
```

Comment:
this is where the magic happens

Reply:
If we used a strategy where we take a `List<uri>` and `function<uri, channel>` we can hold references to fully-wrapped/limited channels, solving the concurrency limiter state problem, and making the code easier to follow without reliance on equals/hashCode in all implementations. For backcompat from Channels.create I think we could pass a random uuid as the uri to retain the old behavior of always-recreate-everything. Thoughts?